### PR TITLE
In-a-box wrapper for localdev testing

### PR DESCRIPTION
**Pull request by dvoytenko**
Related to #7830.

### CODE REVIEW

File: build-system/server-inabox-template.html [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+      name="inabox-atf"
+      src="AD_URL"
+      width="AD_WIDTH" height="AD_HEIGHT"
+      scrolling="no"
+      marginwidth="0"
+      marginheight="0"
+      frameborder="0">
+  </iframe>
+
+  <script>
```
Comment (lannka) [inline build-system/server-inabox-template.html]: do you mind extracting a script file and share with the `inabox.host.html`

File: build-system/server.js [hunk 2]
```diff
@@ -522,13 +522,45 @@ app.use('/a4a(|-3p)/', function(req, res) {
...
+    // `ads.localhost` to ensure that the iframe is fully x-origin.
+    adUrl = urlPrefix.replace('localhost', 'ads.localhost') + adUrl;
+  }
+  const inaboxParam = 'inabox=1';
+  if (adUrl.indexOf('?') == -1) {
+    adUrl += '?' + inaboxParam;
+  } else {
+    adUrl += '&' + inaboxParam;
+  }
+  fs.readFileAsync(process.cwd() + templatePath, 'utf8').then(template => {
```
Comment (lannka) [inline build-system/server.js]: how about inline the HTML here after splitting off the script to a separate file

File: build-system/server-inabox-template.html (hunk 1 above)
Comment (dvoytenko) [inline build-system/server-inabox-template.html]: Can we do this in our `server.js`? I'd definitely prefer if we did that... Or do you mean we use a `<script>` for this?

File: build-system/server.js (hunk 2 above)
Comment (dvoytenko) [inline build-system/server.js]: You mean inline iframe contents?

File: build-system/server-inabox-template.html (hunk 1 above)
Comment (lannka) [inline build-system/server-inabox-template.html]: yes `<script>`

File: build-system/server.js (hunk 2 above)
Comment (lannka) [inline build-system/server.js]: yes

File: contributing/DEVELOPING.md [hunk 3]
```diff
@@ -122,6 +122,26 @@ Additionally, the following query parameters can be provided:
...
 - `width` - the width of the `amp-ad` (default "300")
 - `height` - the height of the `amp-ad` (default "250")
+- `offset` - the offset to push the `amp-ad` down the page (default "0px"). Can be used to push the Ad out of the viewport, e.g. using `offset=150vh`.
+
+
+### In-a-box envelope (/inabox/)
+
+If you are working on AMP In-a-box Ads, you can use the local in-a-box envelope for testing local and production AMP documents with the local JS version.
+
+Make sure to run gulp with `--with_inabox` flag.
```
Comment (zhouyx) [inline contributing/DEVELOPING.md]: can't find where this flag comes from?

Comment (dvoytenko) [inline contributing/DEVELOPING.md]: See here: https://github.com/ampproject/amphtml/blob/a44b83050712361ac7be28db7e9db9686bc64872/gulpfile.js#L986

File: build-system/server-inabox-template.html (hunk 1 above)
Comment (dvoytenko) [inline build-system/server-inabox-template.html]: Done

File: build-system/server.js (hunk 2 above)
Comment (dvoytenko) [inline build-system/server.js]: Discussed offline. Will keep this code.

**Comment by dvoytenko** [comment]
@lannka PTAL

File: examples/inabox.host.html [hunk 4]
```diff
@@ -59,68 +59,6 @@
...
-    }
-
-    function loadScript(win, url, callback) {
-      const script = win.document.createElement('script');
-      script.src = url;
-      script.onload = callback;
-      win.document.body.appendChild(script);
-    }
-  </script>
+  <script src="./inabox.host-bootstrap.js"></script>
```
Comment (lannka) [inline examples/inabox.host.html]: did you check in the new file?
Also, how about calling it: `./inabox-tag-integration.js`

File: build-system/server.js [hunk 5]
```diff
@@ -522,13 +522,45 @@ app.use('/a4a(|-3p)/', function(req, res) {
...
+  var templatePath = '/build-system/server-inabox-template.html';
+  var urlPrefix = getUrlPrefix(req);
+  if (!adUrl.startsWith('/m') &&
+      urlPrefix.indexOf('//localhost') != -1) {
+    // This is a special case for testing. `localhost` URLs are transformed to
+    // `ads.localhost` to ensure that the iframe is fully x-origin.
+    adUrl = urlPrefix.replace('localhost', 'ads.localhost') + adUrl;
+  }
+  const inaboxParam = 'inabox=1';
+  if (adUrl.indexOf('?') == -1) {
```
Comment (lannka) [inline build-system/server.js]: can we extract a method? i saw duplicated code

File: build-system/server.js [hunk 6]
```diff
@@ -522,13 +522,45 @@ app.use('/a4a(|-3p)/', function(req, res) {
...
+// In-a-box envelope.
+// Examples:
+// http://localhost:8000/inabox/examples/animations.amp.max.html
+// http://localhost:8000/inabox/max/s/www.washingtonpost.com/amphtml/news/post-politics/wp/2016/02/21/bernie-sanders-says-lower-turnout-contributed-to-his-nevada-loss-to-hillary-clinton/
+// http://localhost:8000/inabox/min/s/www.washingtonpost.com/amphtml/news/post-politics/wp/2016/02/21/bernie-sanders-says-lower-turnout-contributed-to-his-nevada-loss-to-hillary-clinton/
+app.use('/inabox/', function(req, res) {
+  var adUrl = req.url;
+  var templatePath = '/build-system/server-inabox-template.html';
+  var urlPrefix = getUrlPrefix(req);
+  if (!adUrl.startsWith('/m') &&
```
Comment (lannka) [inline build-system/server.js]: can you document about `/m` ? or be explicitly checking `/min/` and `/max/`

File: build-system/server.js [hunk 7]
```diff
@@ -522,13 +522,45 @@ app.use('/a4a(|-3p)/', function(req, res) {
...
 
+// In-a-box envelope.
+// Examples:
+// http://localhost:8000/inabox/examples/animations.amp.max.html
+// http://localhost:8000/inabox/max/s/www.washingtonpost.com/amphtml/news/post-politics/wp/2016/02/21/bernie-sanders-says-lower-turnout-contributed-to-his-nevada-loss-to-hillary-clinton/
+// http://localhost:8000/inabox/min/s/www.washingtonpost.com/amphtml/news/post-politics/wp/2016/02/21/bernie-sanders-says-lower-turnout-contributed-to-his-nevada-loss-to-hillary-clinton/
+app.use('/inabox/', function(req, res) {
+  var adUrl = req.url;
+  var templatePath = '/build-system/server-inabox-template.html';
+  var urlPrefix = getUrlPrefix(req);
```
Comment (lannka) [inline build-system/server.js]: req.hostname

File: examples/inabox.host.html (hunk 4 above)
Comment (dvoytenko) [inline examples/inabox.host.html]: Done

File: build-system/server.js (hunk 7 above)
Comment (dvoytenko) [inline build-system/server.js]: Why?

File: build-system/server.js (hunk 6 above)
Comment (dvoytenko) [inline build-system/server.js]: Done

File: build-system/server.js (hunk 5 above)
Comment (dvoytenko) [inline build-system/server.js]: Done

**Comment by dvoytenko** [comment]
@lannka PTAL

**Pull request merged**